runtime/frequency/src/lib.rs

```diff
@@ -1235,6 +1249,279 @@ construct_runtime!(
 	}
 );
 
+/// The SignedExtension trait is implemented on StaleHashCheckExtension to validate the
+/// request. The purpose of this is to ensure that the target_hash is verified in transaction pool
+/// before getting into block. This is to reduce the chance of capacity consumption due to stale hash
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+pub struct StaleHashCheckExtension;
+
+/// Extracted data type from transactions to verify target hash
+struct HashCheckData {
+	message_source_id: MessageSourceId,
+	schema_id: SchemaId,
+	page: Option<PageId>,
+	hash: PageHash,
+}
+
+impl HashCheckData {
+	fn new_itemized(
+		message_source_id: MessageSourceId,
+		schema_id: SchemaId,
+		hash: PageHash,
+	) -> Self {
+		Self { message_source_id, schema_id, page: None, hash }
+	}
+
+	fn new_paginated(
+		message_source_id: MessageSourceId,
+		schema_id: SchemaId,
+		page_id: PageId,
+		hash: PageHash,
+	) -> Self {
+		Self { message_source_id, schema_id, page: Some(page_id), hash }
+	}
+}
+
+impl sp_std::fmt::Debug for StaleHashCheckExtension {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "StaleHashCheckExtension")
+	}
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+impl SignedExtension for StaleHashCheckExtension {
```

Comment:
had to bring it out of pallet and inside runtime to be able to support multiple pallets and instead of using generics which was causing issues when used for multiple pallets it's directly using the concrete type used in the runtime

Reply:
Could we have used a similar approach to how we inject CapacityEligibleCalls into the frequency-tx-payment pallet and injected a struct that includes the extract_hash_data function?

Reply:
I think I can make that work. I'll close and merge this and add an issue for that